### DISCUSSION TITLE
Should `company-complete` show lots of completions at beginning of a line (empty input)?

### DISCUSSION THREAD

**Discussion by apc** [company-mode/company-mode#1237] [Q&A] [answered]
Running `company-complete` with the point at the beginning of a line, when in `sh-mode`, triggers a long list of completion options. Is this by design? Is there a way of forcing `company-complete` to ignore the empty string?

I ended up running into trouble when using `company-complete` in `sh-mode` with `tab-always-indent` set to `'complete`. Trying to press `tab` at the beginning of a line leads to a somewhat useless list of completion options. This doesn't happen when in e.g. `org-mode`---I presume this is because when I run `company-complete` on the empty string in `org-mode` this results in a `No completion found`message. 

The following seems to reproduce the behavior I'm talking about. 

- Start vanilla emacs: `emacs -q`
- Manually install `company` and call `company-mode`.
- Enter `sh-mode` in an empty buffer.
- Eval `(setq tab-always-indent 'complete)`.
- Press TAB at the beginning of a line.

**Comment by dgutov** [comment, marked as answer]
It's absolutely intended: if you ask for completion on an empty prefix, it will give you whatever completions it has. Why not?

Your recipe seems to be missing some binding, I guess. How do you get `TAB` to call `company-complete`?

> I presume this is because when I run company-complete on the empty string in org-mode this results in a No completion foundmessage

In the end, though, the behavior depends on whatever a backend returns. If Org's completion rejects empty strings, we can't force it to do otherwise.

**Comment by apc** [comment]
Thanks so much for your reply. I suspected as much, but was trying to understand the cause of some undesired behavior. I will close this issue, since I no longer think it's related to `company-mode`. Sorry for taking your time!

**Comment by dgutov** [comment]
Other options are:

* Adding an around-advice on `sh-completion-at-point-function` to return nil when the point is at bol, or the input is empty, etc.
* Adding a new element to `company-transformers` which would look at `company-prefix`, see it's an empty string, and return nil.
* Changing the default value of `company-candidates-predicate` to do the same.

Not sure which one to recommend; try 1 or 2. Number 3 seems like the most risky one, though who knows.

Keep in mind, though, that you might want to show completions for empty prefix when using other completion backends/sources. If so, number 1 is your best bet.
